File: tests/polyglot-pilot/go/common/statedelta/state_delta_test.go
// state_delta_test.go — rapid + standard testing contract tests for the Go port.
//
// Mirrors `tests/state_delta/unit/test_matcher.py` and the predicate tests for
// the Python canonical. Property-based assertions via pgregory.net/rapid where
// the universe semantic is quantified (one PBT per predicate is the
// layered-unit minimum).

package statedelta

import (
	"errors"
	"sort"
	"testing"

	"pgregory.net/rapid"
)

// ---------------------------------------------------------------------------
// Walking skeleton + core semantics
// ---------------------------------------------------------------------------

func TestAssertStateDelta_ReturnsOnCleanPrependTransition(t *testing.T) {
	AssertStateDelta(t,
		Snapshot{"PATH": "/usr/bin"},
		Snapshot{"PATH": "/des/bin:/usr/bin"},
		[]string{"PATH"},
		map[string]Predicate{"PATH": PrependedWith("/des/bin")},
	)
}

func TestAssertStateDeltaErr_ReportsPredicateFailureWithFullContext(t *testing.T) {
	err := AssertStateDeltaErr(
		Snapshot{"PATH": "/usr/bin"},
		Snapshot{"PATH": "/wrong"},
		[]string{"PATH"},
		map[string]Predicate{"PATH": PrependedWith("/des/bin")},
	)
	if err == nil {
		t.Fatal("expected violation, got nil")
	}
	var sdv *StateDeltaViolation
	if !errors.As(err, &sdv) {
		t.Fatalf("expected *StateDeltaViolation, got %T", err)
	}
	if len(sdv.Violations) != 1 {
		t.Fatalf("expected 1 violation, got %d", len(sdv.Violations))
	}
	v := sdv.Violations[0]
	if v.Kind != PredicateFailed {
		t.Errorf("kind: want %q got %q", PredicateFailed, v.Kind)
	}
	if v.Key != "PATH" {
		t.Errorf("key: want PATH got %q", v.Key)
	}
	if v.PredicateName != "prepended_with(/des/bin)" {
		t.Errorf("predicate name: want prepended_with(/des/bin) got %q", v.PredicateName)
	}
}

func TestAssertStateDeltaErr_ImplicitUnchangedCatchesUndeclaredChange(t *testing.T) {
	err := AssertStateDeltaErr(
		Snapshot{"PATH": "/u/bin", "HOME": "/home/u"},
		Snapshot{"PATH": "/des/bin:/u/bin", "HOME": "/home/changed"},
		[]string{"PATH", "HOME"},
		map[string]Predicate{"PATH": PrependedWith("/des/bin")},
	)
	if err == nil {
		t.Fatal("expected violation, got nil")
	}
	sdv := err.(*StateDeltaViolation)
	if len(sdv.Violations) != 1 {
		t.Fatalf("expected 1 violation, got %d", len(sdv.Violations))
	}
	v := sdv.Violations[0]
	if v.Kind != UndeclaredChange || v.Key != "HOME" {
		t.Errorf("expected UndeclaredChange on HOME, got %+v", v)
	}
}

func TestAssertStateDeltaErr_CollectsMultipleViolationsA7(t *testing.T) {
	err := AssertStateDeltaErr(
		Snapshot{"PATH": "/u", "HOME": "/h", "X": "1"},
		Snapshot{"PATH": "/wrong", "HOME": "/h2", "X": "1"},
		[]string{"PATH", "HOME", "X"},
		map[string]Predicate{
			"PATH": PrependedWith("/des"),
			"HOME": Unchanged(),
		},
	)
	if err == nil {
		t.Fatal("expected violations, got nil")
	}
	sdv := err.(*StateDeltaViolation)
	if len(sdv.Violations) != 2 {
		t.Fatalf("expected 2 violations, got %d", len(sdv.Violations))
	}
	keys := []string{sdv.Violations[0].Key, sdv.Violations[1].Key}
	sort.Strings(keys)
	if keys[0] != "HOME" || keys[1] != "PATH" {
		t.Errorf("expected keys [HOME PATH], got %v", keys)
	}
}

func TestAssertStateDeltaErr_StrictModeFlagsExtraKeys(t *testing.T) {
	err := AssertStateDeltaErr(
		Snapshot{"PATH": "/u", "EXTRA": "x"},
		Snapshot{"PATH": "/des:/u", "EXTRA": "x2"},
		[]string{"PATH"},
		map[string]Predicate{"PATH": PrependedWith("/des")},
		Options{Strict: true},
	)
	if err == nil {
		t.Fatal("expected violation, got nil")
	}
	sdv := err.(*StateDeltaViolation)
	found := false
	for _, v := range sdv.Violations {
		if v.Kind == StrictUniverseMismatch && v.Key == "EXTRA" {
			found = true
		}
	}
	if !found {
		t.Errorf("expected StrictUniverseMismatch for EXTRA, got %+v", sdv.Violations)
	}
}

// ---------------------------------------------------------------------------
// Predicate library — Python parity
// ---------------------------------------------------------------------------

func TestUnchanged_PassesIffDeepEqual(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		v := rapid.IntRange(-1000, 1000).Draw(t, "v")
		if !Unchanged().Eval(v, v).Ok {
			t.Fatalf("unchanged should pass for identical value %d", v)
		}
	})
	if Unchanged().Eval(1, 2).Ok {
		t.Error("unchanged should fail for 1 vs 2")
	}
	if !Unchanged().Eval(map[string]int{"a": 1}, map[string]int{"a": 1}).Ok {
		t.Error("unchanged should pass for deep-equal maps")
	}
}

func TestPrependedWith_StringComposition(t *testing.T) {
	if !PrependedWith("/des/bin").Eval("/usr/bin", "/des/bin:/usr/bin").Ok {
		t.Error("expected pass for clean prepend")
	}
	if PrependedWith("/des/bin").Eval("/usr/bin", "/wrong").Ok {
		t.Error("expected fail for wrong composition")
	}
	rapid.Check(t, func(t *rapid.T) {
		tail := rapid.String().Draw(t, "tail")
		composed := "PRE:" + tail
		if !PrependedWith("PRE").Eval(tail, composed).Ok {
			t.Fatalf("PBT: expected PRE:%q composition to pass", tail)
		}
	})
}

func TestAppendedWith_StringComposition(t *testing.T) {
	if !AppendedWith(".bak").Eval("/etc/hosts", "/etc/hosts:.bak").Ok {
		t.Error("expected pass for clean append")
	}
	if AppendedWith(".bak").Eval("/etc/hosts", "/etc/hosts").Ok {
		t.Error("expected fail for missing suffix")
	}
}

func TestSetTo_IgnoresOldMatchesNew(t *testing.T) {
	if !SetTo("active").Eval("inactive", "active").Ok {
		t.Error("set_to should ignore old")
	}
	if SetTo("active").Eval("inactive", "pending").Ok {
		t.Error("set_to should fail when new differs")
	}
	if !SetTo(map[string]int{"k": 1}).Eval(nil, map[string]int{"k": 1}).Ok {
		t.Error("set_to should deep-equal complex values")
	}
}

func TestContaining_StringAndSlice(t *testing.T) {
	if !Containing("/usr/bin").Eval("", "/des/bin:/usr/bin").Ok {
		t.Error("containing should find substring")
	}
	if Containing("/usr/bin").Eval("", "/des/bin:/opt/bin").Ok {
		t.Error("containing should fail when substring absent")
	}
	if !Containing(map[string]int{"id": 1}).Eval(nil, []interface{}{
		map[string]int{"id": 1},
		map[string]int{"id": 2},
	}).Ok {
		t.Error("containing should find deep-equal element in slice")
	}
}

func TestNormalizedTo_ComparesUnderNormaliser(t *testing.T) {
	expandHome := func(v interface{}) interface{} {
		if s, ok := v.(string); ok {
			// trivial $HOME expansion
			for i := 0; i+5 <= len(s); i++ {
				if s[i:i+5] == "$HOME" {
					return s[:i] + "/home/u" + s[i+5:]
				}
			}
		}
		return v
	}
	if !NormalizedTo(expandHome).Eval("/home/u/.local/bin", "$HOME/.local/bin").Ok {
		t.Error("normalized_to should equate $HOME-expanded paths")
	}
	if NormalizedTo(expandHome).Eval("/home/u/.local/bin", "$HOME/.other/bin").Ok {
		t.Error("normalized_to should fail when normalised values differ")
	}
}

func TestIdempotentAfter_FirstSegmentMatches(t *testing.T) {
	if !IdempotentAfter("DES_BIN").Eval("anything", "DES_BIN:/usr/bin").Ok {
		t.Error("idempotent_after should pass when prefix is first segment")
	}
	if IdempotentAfter("DES_BIN").Eval("anything", "/usr/bin:/opt/bin").Ok {
		t.Error("idempotent_after should fail when first segment differs")
	}
}

func TestLegacyHealed_FourCasePaperTrace(t *testing.T) {
	const legacy = "DES_BIN:SYSTEM_PATH_FALLBACK"
	detector := func(v interface{}) bool { s, _ := v.(string); return s == legacy }
	healed := func(v interface{}) bool {
		s, _ := v.(string)
		return s != legacy && len(s) > 8 && s[:8] == "DES_BIN:"
	}
	pred := LegacyHealed(detector, healed)
	if !pred.Eval(legacy, "DES_BIN:/usr/bin").Ok {
		t.Error("legacy_healed should pass when old is legacy and new is healed")
	}
	if pred.Eval(legacy, legacy).Ok {
		t.Error("legacy_healed should fail when new is still legacy")
	}
	if pred.Eval("/usr/bin", "DES_BIN:/usr/bin").Ok {
		t.Error("legacy_healed should fail when old was not legacy")
	}
}

func TestPrependedWithItem_SliceComposition(t *testing.T) {
	if !PrependedWithItem("a").Eval([]interface{}{}, []interface{}{"a"}).Ok {
		t.Error("expected pass for prepend to empty slice")
	}
	if !PrependedWithItem("a").Eval([]interface{}{"x"}, []interface{}{"a", "x"}).Ok {
		t.Error("expected pass for prepend")
	}
	if PrependedWithItem("a").Eval([]interface{}{"x"}, []interface{}{"x", "a"}).Ok {
		t.Error("expected fail when item is appended not prepended")
	}
	if PrependedWithItem("a").Eval([]interface{}{"x"}, []interface{}{"a"}).Ok {
		t.Error("expected fail when tail is lost")
	}
}

func TestAppendedWithItem_SliceComposition(t *testing.T) {
	if !AppendedWithItem("z").Eval([]interface{}{}, []interface{}{"z"}).Ok {
		t.Error("expected pass for append to empty slice")
	}
	if !AppendedWithItem("z").Eval([]interface{}{"x"}, []interface{}{"x", "z"}).Ok {
		t.Error("expected pass for append")
	}
	if AppendedWithItem("z").Eval([]interface{}{"x"}, []interface{}{"z", "x"}).Ok {
		t.Error("expected fail when item is prepended not appended")
	}
}

// ---------------------------------------------------------------------------
// Universe semantics — PBT
// ---------------------------------------------------------------------------

func TestUniverse_ForbidsHiddenMutationOnAdjacentSlot(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		oldHome := rapid.StringMatching(`[a-z]{1,8}`).Draw(t, "oldHome")
		newHome := rapid.StringMatching(`[a-z]{1,8}`).Draw(t, "newHome")
		if oldHome == newHome {
			t.Skip("trivial — equal values would not trigger violation")
		}
		err := AssertStateDeltaErr(
			Snapshot{"PATH": "/u", "HOME": oldHome},
			Snapshot{"PATH": "/des:/u", "HOME": newHome},
			[]string{"PATH", "HOME"},
			map[string]Predicate{"PATH": PrependedWith("/des")},
		)
		if err == nil {
			t.Fatalf("expected violation for hidden HOME mutation %q -> %q", oldHome, newHome)
		}
	})
}

func TestUniverse_PermitsMutationOnlyWithMatchingPredicate(t *testing.T) {
	err := AssertStateDeltaErr(
		Snapshot{"PATH": "/u", "HOME": "/h"},
		Snapshot{"PATH": "/des:/u", "HOME": "/h"},
		[]string{"PATH", "HOME"},
		map[string]Predicate{"PATH": PrependedWith("/des")},
	)
	if err != nil {
		t.Fatalf("expected no violation, got %v", err)
	}
}

File: tests/polyglot-pilot/cpp/common/statedelta/state_delta_test.cpp
// state_delta_test.cpp — contract tests for the C++ state-delta port.
//
// Mirrors `tests/state_delta/unit/test_matcher.py` and the Go/C# ports'
// predicate tests. No PBT library is available in this pilot's build
// environment (zero third-party test-framework dependency by design) — the
// "property" tests below are small hand-rolled bounded loops over generated
// inputs, not a substitute for `nwave_ai.state_delta`'s Hypothesis suite.

#include <random>
#include <string>

#include "../testkit/testkit.hpp"
#include "state_delta.hpp"

using namespace statedelta;

namespace {

std::string random_lowercase_string(std::mt19937& rng, int min_len, int max_len) {
    std::uniform_int_distribution<int> len_dist(min_len, max_len);
    std::uniform_int_distribution<int> char_dist('a', 'z');
    int len = len_dist(rng);
    std::string s;
    for (int i = 0; i < len; ++i) {
        s.push_back(static_cast<char>(char_dist(rng)));
    }
    return s;
}

} // namespace

// ---------------------------------------------------------------------------
// Walking skeleton + core semantics
// ---------------------------------------------------------------------------

TEST("assert_state_delta returns cleanly on a clean prepend transition") {
    assert_state_delta(
        Snapshot{{"PATH", Value{"/usr/bin"}}},
        Snapshot{{"PATH", Value{"/des/bin:/usr/bin"}}},
        {"PATH"},
        {{"PATH", prepended_with("/des/bin")}});
}

TEST("assert_state_delta_collect reports predicate failure with full context") {
    std::vector<Violation> violations = assert_state_delta_collect(
        Snapshot{{"PATH", Value{"/usr/bin"}}},
        Snapshot{{"PATH", Value{"/wrong"}}},
        {"PATH"},
        {{"PATH", prepended_with("/des/bin")}});
    ASSERT_TRUE(violations.size() == 1);
    const Violation& v = violations[0];
    ASSERT_TRUE(v.kind == ViolationKind::PredicateFailed);
    ASSERT_TRUE(v.key == "PATH");
    ASSERT_TRUE(v.predicate_name == "prepended_with(/des/bin)");
}

TEST("implicit-unchanged enforcement catches an undeclared change") {
    std::vector<Violation> violations = assert_state_delta_collect(
        Snapshot{{"PATH", Value{"/u/bin"}}, {"HOME", Value{"/home/u"}}},
        Snapshot{{"PATH", Value{"/des/bin:/u/bin"}}, {"HOME", Value{"/home/changed"}}},
        {"PATH", "HOME"},
        {{"PATH", prepended_with("/des/bin")}});
    ASSERT_TRUE(violations.size() == 1);
    ASSERT_TRUE(violations[0].kind == ViolationKind::UndeclaredChange);
    ASSERT_TRUE(violations[0].key == "HOME");
}

TEST("assert_state_delta_collect aggregates multiple violations (A7)") {
    std::vector<Violation> violations = assert_state_delta_collect(
        Snapshot{{"PATH", Value{"/u"}}, {"HOME", Value{"/h"}}, {"X", Value{"1"}}},
        Snapshot{{"PATH", Value{"/wrong"}}, {"HOME", Value{"/h2"}}, {"X", Value{"1"}}},
        {"PATH", "HOME", "X"},
        {{"PATH", prepended_with("/des")}, {"HOME", unchanged()}});
    ASSERT_TRUE(violations.size() == 2);
    ASSERT_TRUE((violations[0].key == "HOME" && violations[1].key == "PATH") ||
                (violations[0].key == "PATH" && violations[1].key == "HOME"));
}

TEST("strict mode flags keys present outside the declared universe") {
    std::vector<Violation> violations = assert_state_delta_collect(
        Snapshot{{"PATH", Value{"/u"}}, {"EXTRA", Value{"x"}}},
        Snapshot{{"PATH", Value{"/des:/u"}}, {"EXTRA", Value{"x2"}}},
        {"PATH"},
        {{"PATH", prepended_with("/des")}},
        AssertOptions{/*strict=*/true});
    bool found = false;
    for (const auto& v : violations) {
        if (v.kind == ViolationKind::StrictUniverseMismatch && v.key == "EXTRA") {
            found = true;
        }
    }
    ASSERT_TRUE(found);
}

// ---------------------------------------------------------------------------
// Predicate library — parity with the Python/Go/C# canonical
// ---------------------------------------------------------------------------

TEST("unchanged() passes iff deep-equal") {
    std::mt19937 rng(42);
    for (int i = 0; i < 30; ++i) {
        std::string v = random_lowercase_string(rng, 0, 12);
        ASSERT_TRUE(unchanged().eval(Value{v}, Value{v}).ok);
    }
    ASSERT_TRUE(!unchanged().eval(Value{"a"}, Value{"b"}).ok);
    Record r1{{"email", "a@x.com"}};
    Record r2{{"email", "a@x.com"}};
    ASSERT_TRUE(unchanged().eval(Value{std::vector<Record>{r1}}, Value{std::vector<Record>{r2}}).ok);
}

TEST("prepended_with composes old and new by string concatenation") {
    ASSERT_TRUE(prepended_with("/des/bin").eval(Value{"/usr/bin"}, Value{"/des/bin:/usr/bin"}).ok);
    ASSERT_TRUE(!prepended_with("/des/bin").eval(Value{"/usr/bin"}, Value{"/wrong"}).ok);

    std::mt19937 rng(7);
    for (int i = 0; i < 30; ++i) {
        std::string tail = random_lowercase_string(rng, 0, 10);
        std::string composed = "PRE:" + tail;
        ASSERT_TRUE(prepended_with("PRE").eval(Value{tail}, Value{composed}).ok);
    }
}

TEST("appended_with composes old and new by string concatenation") {
    ASSERT_TRUE(appended_with(".bak").eval(Value{"/etc/hosts"}, Value{"/etc/hosts:.bak"}).ok);
    ASSERT_TRUE(!appended_with(".bak").eval(Value{"/etc/hosts"}, Value{"/etc/hosts"}).ok);
}

TEST("set_to ignores old and compares new by value") {
    ASSERT_TRUE(set_to(Value{"active"}).eval(Value{"inactive"}, Value{"active"}).ok);
    ASSERT_TRUE(!set_to(Value{"active"}).eval(Value{"inactive"}, Value{"pending"}).ok);
}

TEST("containing finds substrings and list membership") {
    ASSERT_TRUE(containing(Value{"/usr/bin"}).eval(Value{""}, Value{"/des/bin:/usr/bin"}).ok);
    ASSERT_TRUE(!containing(Value{"/usr/bin"}).eval(Value{""}, Value{"/des/bin:/opt/bin"}).ok);

    Record target{{"id", "1"}};
    Record other{{"id", "2"}};
    ASSERT_TRUE(containing(Value{std::vector<Record>{target}})
                    .eval(Value{}, Value{std::vector<Record>{other, target}})
                    .ok);
}

TEST("normalized_to compares under a normaliser") {
    auto expand_home = [](const Value& v) -> Value {
        if (!v.is_string()) return v;
        const std::string& s = v.as_string();
        std::string needle = "$HOME";
        auto pos = s.find(needle);
        if (pos == std::string::npos) return v;
        return Value{s.substr(0, pos) + "/home/u" + s.substr(pos + needle.size())};
    };
    ASSERT_TRUE(normalized_to(expand_home)
                    .eval(Value{"/home/u/.local/bin"}, Value{"$HOME/.local/bin"})
                    .ok);
    ASSERT_TRUE(!normalized_to(expand_home)
                     .eval(Value{"/home/u/.local/bin"}, Value{"$HOME/.other/bin"})
                     .ok);
}

TEST("idempotent_after passes when the first segment already matches") {
    ASSERT_TRUE(idempotent_after("DES_BIN").eval(Value{"anything"}, Value{"DES_BIN:/usr/bin"}).ok);
    ASSERT_TRUE(!idempotent_after("DES_BIN").eval(Value{"anything"}, Value{"/usr/bin:/opt/bin"}).ok);
}

TEST("legacy_healed follows the four-case paper trace") {
    const std::string legacy = "DES_BIN:SYSTEM_PATH_FALLBACK";
    auto detector = [legacy](const Value& v) { return v.is_string() && v.as_string() == legacy; };
    auto healed = [legacy](const Value& v) {
        if (!v.is_string()) return false;
        const std::string& s = v.as_string();
        return s != legacy && s.size() > 8 && s.substr(0, 8) == "DES_BIN:";
    };
    Predicate pred = legacy_healed(detector, healed);
    ASSERT_TRUE(pred.eval(Value{legacy}, Value{"DES_BIN:/usr/bin"}).ok);
    ASSERT_TRUE(!pred.eval(Value{legacy}, Value{legacy}).ok);
    ASSERT_TRUE(!pred.eval(Value{"/usr/bin"}, Value{"DES_BIN:/usr/bin"}).ok);
}

TEST("prepended_with_item composes list state by prepending one item") {
    Record a{{"v", "a"}};
    Record x{{"v", "x"}};
    ASSERT_TRUE(prepended_with_item(a).eval(Value{std::vector<Record>{}}, Value{std::vector<Record>{a}}).ok);
    ASSERT_TRUE(
        prepended_with_item(a).eval(Value{std::vector<Record>{x}}, Value{std::vector<Record>{a, x}}).ok);
    ASSERT_TRUE(
        !prepended_with_item(a).eval(Value{std::vector<Record>{x}}, Value{std::vector<Record>{x, a}}).ok);
    ASSERT_TRUE(
        !prepended_with_item(a).eval(Value{std::vector<Record>{x}}, Value{std::vector<Record>{a}}).ok);
}

TEST("appended_with_item composes list state by appending one item") {
    Record z{{"v", "z"}};
    Record x{{"v", "x"}};
    ASSERT_TRUE(appended_with_item(z).eval(Value{std::vector<Record>{}}, Value{std::vector<Record>{z}}).ok);
    ASSERT_TRUE(
        appended_with_item(z).eval(Value{std::vector<Record>{x}}, Value{std::vector<Record>{x, z}}).ok);
    ASSERT_TRUE(
        !appended_with_item(z).eval(Value{std::vector<Record>{x}}, Value{std::vector<Record>{z, x}}).ok);
}

// ---------------------------------------------------------------------------
// Universe semantics — bounded hand-rolled property loops
// ---------------------------------------------------------------------------

TEST("universe forbids a hidden mutation on an adjacent slot") {
    std::mt19937 rng(99);
    for (int i = 0; i < 20; ++i) {
        std::string old_home = random_lowercase_string(rng, 1, 8);
        std::string new_home = random_lowercase_string(rng, 1, 8);
        if (old_home == new_home) continue;
        std::vector<Violation> violations = assert_state_delta_collect(
            Snapshot{{"PATH", Value{"/u"}}, {"HOME", Value{old_home}}},
            Snapshot{{"PATH", Value{"/des:/u"}}, {"HOME", Value{new_home}}},
            {"PATH", "HOME"},
            {{"PATH", prepended_with("/des")}});
        ASSERT_TRUE(!violations.empty());
    }
}

TEST("universe permits mutation only when a matching predicate is declared") {
    std::vector<Violation> violations = assert_state_delta_collect(
        Snapshot{{"PATH", Value{"/u"}}, {"HOME", Value{"/h"}}},
        Snapshot{{"PATH", Value{"/des:/u"}}, {"HOME", Value{"/h"}}},
        {"PATH", "HOME"},
        {{"PATH", prepended_with("/des")}});
    ASSERT_TRUE(violations.empty());
}
